snapshotter/demux/metrics/proxy.go

```diff
@@ -0,0 +1,144 @@
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/containerd/containerd/log"
+)
+
+const (
+	// RequestPath is the path in the microVM hosting remote snapshotter metrics.
+	RequestPath = "http://localhost/metrics"
+)
+
+// Proxy represents a metrics proxy server for getting and serving remote snapshotter metrics.
+type Proxy struct {
+	// Client is the HTTP client used to send metrics requests to a remote snapshotter.
+	Client *http.Client
+	// Server is the proxy server on host for serving remote snapshotter metrics.
+	Server *http.Server
+	// Port is the port that the metrics proxy server listens on.
+	Port int
+}
+
+// NewProxy creates a new Proxy with initialized HTTP client and port.
+// dialer is used as the DialContext underlying the metrics HTTP client's RoundTripper.
+//
+// Reference: https://pkg.go.dev/net/http#Transport
+func NewProxy(portRange string, dialer func(context.Context, string, string) (net.Conn, error)) (*Proxy, error) {
+	port, err := findOpenPort(portRange)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Proxy{
+		Client: &http.Client{
+			Transport: &http.Transport{
+				DialContext: dialer,
+			},
+		},
+		Server: nil,
+		Port:   port,
+	}, nil
+}
+
+// findOpenPort uses net.Listen to determine if a port is available for hosting a metrics proxy server.
+func findOpenPort(portRange string) (int, error) {
+	ports := strings.Split(portRange, "-")
+	portRangeError := fmt.Errorf("invalid port range %s", portRange)
+	if len(ports) < 2 {
+		return 0, portRangeError
+	}
+	lower, err := strconv.Atoi(ports[0])
+	if err != nil {
+		return 0, portRangeError
+	}
+	upper, err := strconv.Atoi(ports[1])
+	if err != nil {
+		return 0, portRangeError
+	}
+
+	for port := lower; port <= upper; port++ {
+		ln, err := net.Listen("tcp", ":"+strconv.Itoa(port))
+		if err != nil {
+			continue
+		}
+		ln.Close()
+		if err != nil {
+			continue
+		}
+		return port, nil
+	}
+
+	return 0, fmt.Errorf("could not find open port in range %s", portRange)
+}
+
+// Serve starts the metrics proxy server for a single in-VM snapshotter.
+func (mp *Proxy) Serve(ctx context.Context) error {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/metrics", mp.metrics)
+
+	mp.Server = &http.Server{
+		Addr:    ":" + strconv.Itoa(mp.Port),
+		Handler: mux,
+	}
+
+	err := mp.Server.ListenAndServe()
+	if err != http.ErrServerClosed {
+		return err
+	}
+
+	return nil
+}
+
+// Shutdown shuts down the remote snapshotter metrics proxy server.
+func (mp *Proxy) Shutdown(ctx context.Context) error {
+	return mp.Server.Shutdown(ctx)
+}
+
+// metrics is an HTTP handler receiving requests on /metrics for a given VM/snapshotter.
+func (mp *Proxy) metrics(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	rawMetrics, err := mp.pullVMSnapshotterMetrics(ctx)
+	if err != nil {
+		log.G(ctx).Errorf("error reading response body: %v\n", err)
+	}
+	// Simply write the metrics string to ResponseWriter
+	fmt.Fprintf(w, "%s", rawMetrics)
```

Comment:
Hmm. Maybe we should use `io.Copy` to stream the response body to the `ResponseWriter` rather than buffering the full response in memory.

Reply:
Ah yes good callout, can update.